File: apps/vscode/src/services/devops-mcp/server/providers/github.ts
/** GitHub (github.com and GitHub Enterprise Server) through the REST API. */
import { type Auth, githubAuth } from "../auth"
import { DevOpsError } from "../errors"
import type { Remote } from "../repo"
import {
	type Check,
	type FailedStep,
	type Fetch,
	Http,
	type JobResult,
	type PipelineRun,
	type Provider,
	type PullRequest,
	type Result,
	type RunReport,
	type Status,
	tail,
} from "./types"

const TIMESTAMP = /^\d{4}-\d\d-\d\dT[\d:.]+Z /gm

const RESULTS: Record<string, string> = {
	success: "success",
	failure: "failure",
	timed_out: "failure",
	startup_failure: "failure",
	cancelled: "cancelled",
	skipped: "skipped",
	neutral: "skipped",
	stale: "skipped",
	action_required: "action_required",
	error: "failure", // commit status API
}

const result = (value?: string | null): Result => (value ? (RESULTS[value] ?? value) : undefined)

const status = (value?: string | null): Status =>
	value === "completed" ? "completed" : value === "in_progress" ? "in_progress" : "queued"

/**
 * The `lines` lines ending at the last `##[error]` line of a job log.
 *
 * GitHub serves one log per job, and its tail is post-job cleanup; the useful
 * output is what the failing step printed just before the runner recorded the error.
 */
export function failureExcerpt(log: string, lines: number): string {
	const all = log.trimEnd().split(/\r?\n/)
	let last = -1
	all.forEach((line, i) => {
		if (line.includes("##[error]")) {
			last = i
		}
	})
	if (last < 0) {
		return tail(log, lines)
	}
	return all.slice(Math.max(0, last + 1 - lines), last + 1).join("\n")
}

export class GitHubProvider implements Provider {
	readonly kind = "GitHub"
	readonly maxBodyLength = 65536
	readonly repoUrl: string
	private readonly api: string
	private readonly http: Http

	constructor(
		private readonly remote: Remote,
		fetchImpl?: Fetch,
		auth: Auth = githubAuth(remote.host),
	) {
		const base = remote.host === "github.com" ? "https://api.github.com" : `https://${remote.host}/api/v3`
		this.api = `${base}/repos/${remote.owner}/${remote.repo}`
		this.repoUrl = `https://${remote.host}/${remote.owner}/${remote.repo}`
		this.http = new Http(auth, { Accept: "application/vnd.github+json", "X-GitHub-Api-Version": "2022-11-28" }, fetchImpl)
	}

	private get<T = any>(path: string, query?: Record<string, string | number | undefined>): Promise<T> {
		return this.http.json<T>("GET", `${this.api}${path}`, { query })
	}

	private pr(d: any): PullRequest {
		return {
			id: d.number,
			title: d.title,
			body: d.body ?? "",
			state: d.merged_at ? "merged" : d.state,
			draft: Boolean(d.draft),
			sourceBranch: d.head.ref,
			targetBranch: d.base.ref,
			author: d.user?.login ?? "",
			url: d.html_url,
			headSha: d.head.sha,
		}
	}

	private run(d: any): PipelineRun {
		return {
			id: d.id,
			name: d.name || d.display_title || String(d.id),
			status: status(d.status),
			result: result(d.conclusion),
			branch: d.head_branch ?? "",
			commit: d.head_sha,
			url: d.html_url,
			started: d.run_started_at ?? d.created_at,
			finished: d.status === "completed" ? d.updated_at : undefined,
			event: d.event,
		}
	}

	async defaultBranch(): Promise<string> {
		return (await this.get("")).default_branch
	}

	async findOpenPr(branch: string): Promise<PullRequest | undefined> {
		const prs = await this.get<any[]>("/pulls", { head: `${this.remote.owner}:${branch}`, state: "open" })
		return prs.length ? this.pr(prs[0]) : undefined
	}

	async getPr(id: number): Promise<PullRequest> {
		return this.pr(await this.get(`/pulls/${id}`))
	}

	async createPr(title: string, body: string, source: string, target: string, draft: boolean): Promise<PullRequest> {
		const json = { title, body, head: source, base: target, draft }
		return this.pr(await this.http.json("POST", `${this.api}/pulls`, { json }))
	}

	async updatePr(id: number, title: string | undefined, body: string | undefined): Promise<PullRequest> {
		const json: Record<string, string> = {}
		if (title !== undefined) json.title = title
		if (body !== undefined) json.body = body
		return this.pr(await this.http.json("PATCH", `${this.api}/pulls/${id}`, { json }))
	}

	async listRuns(branch: string | undefined, pr: PullRequest | undefined, limit: number): Promise<PipelineRun[]> {
		const query: Record<string, string | number | undefined> = { per_page: limit }
		if (pr?.headSha) {
			query.head_sha = pr.headSha
		} else if (branch) {
			query.branch = branch
		}
		const data = await this.get("/actions/runs", query)
		return (data.workflow_runs ?? []).slice(0, limit).map((r: any) => this.run(r))
	}

	async runReport(runId: number, logLines: number): Promise<RunReport> {
		const run = this.run(await this.get(`/actions/runs/${runId}`))
		const jobsData: any[] = (await this.get(`/actions/runs/${runId}/jobs`, { filter: "latest", per_page: 100 })).jobs ?? []
		const jobs: JobResult[] = jobsData.map((j) => ({
			name: j.name,
			status: status(j.status),
			result: result(j.conclusion),
			url: j.html_url,
		}))
		const failures: FailedStep[] = []
		for (const j of jobsData) {
			if (result(j.conclusion) !== "failure") {
				continue
			}
			const steps = (j.steps ?? []).filter((s: any) => result(s.conclusion) === "failure").map((s: any) => s.name)
			failures.push({
				job: j.name,
				step: steps.join(", ") || "(job)",
				errors: await this.annotations(j.id),
				logTail: await this.jobLog(j.id, logLines),
			})
		}
		return { run, jobs, failures }
	}

	private async annotations(jobId: number): Promise<string[]> {
		// A job's id is also its check-run id; failure annotations carry the error messages.
		try {
			const notes = await this.get<any[]>(`/check-runs/${jobId}/annotations`, { per_page: 50 })
			return notes
				.filter((n) => n.annotation_level === "failure")
				.map((n) => n.message)
				.slice(0, 20)
		} catch (error) {
			if (error instanceof DevOpsError) return []
			throw error
		}
	}

	private async jobLog(jobId: number, lines: number): Promise<string | undefined> {
		if (lines <= 0) {
			return undefined
		}
		try {
			const resp = await this.http.request("GET", `${this.api}/actions/jobs/${jobId}/logs`)
			return failureExcerpt((await resp.text()).replace(TIMESTAMP, ""), lines)
		} catch (error) {
			// Logs expire or are not ready yet; the report is still useful without them.
			if (error instanceof DevOpsError) return undefined
			throw error
		}
	}

	async prChecks(pr: PullRequest): Promise<Check[]> {
		const sha = pr.headSha ?? (await this.getPr(pr.id)).headSha
		const runs: any[] = (await this.get(`/commits/${sha}/check-runs`, { per_page: 100 })).check_runs ?? []
		const checks: Check[] = runs.map((r) => ({
			name: r.name,
			status: status(r.status),
			result: result(r.conclusion),
			url: r.html_url,
		}))
		const statuses: any[] = (await this.get(`/commits/${sha}/status`)).statuses ?? []
		for (const s of statuses) {
			const done = s.state !== "pending"
			checks.push({
				name: s.context,
				status: done ? "completed" : "in_progress",
				result: done ? result(s.state) : undefined,
				url: s.target_url ?? undefined,
			})
		}
		return checks
	}
}

File: tools/devops-mcp/src/devops_mcp/providers/github.py
"""GitHub (github.com and GitHub Enterprise Server) via the REST API."""

from __future__ import annotations

import re
from typing import Any

import httpx

from ..auth import GitHubAuth
from ..errors import DevOpsError
from ..repo import Remote
from .base import (
    Check,
    FailedStep,
    Http,
    JobResult,
    PipelineRun,
    Provider,
    PullRequest,
    RunReport,
    tail,
)

_TIMESTAMP = re.compile(r"^\d{4}-\d\d-\d\dT[\d:.]+Z ", re.MULTILINE)

_RESULTS = {
    "success": "success",
    "failure": "failure",
    "timed_out": "failure",
    "startup_failure": "failure",
    "cancelled": "cancelled",
    "skipped": "skipped",
    "neutral": "skipped",
    "stale": "skipped",
    "action_required": "action_required",
    "error": "failure",  # commit status API
}


def failure_excerpt(log: str, lines: int) -> str:
    """The `lines` lines ending at the last `##[error]` line of a job log.

    GitHub serves one log per job, and its tail is post-job cleanup; the useful output
    is what the failing step printed just before the runner recorded the error.
    """
    all_lines = log.rstrip().splitlines()
    errors = [i for i, line in enumerate(all_lines) if "##[error]" in line]
    if not errors:
        return tail(log, lines)
    end = errors[-1] + 1
    return "\n".join(all_lines[max(0, end - lines) : end])


def _result(value: str | None) -> str | None:
    return _RESULTS.get(value, value) if value else None


def _status(value: str | None) -> str:
    if value == "completed":
        return "completed"
    if value == "in_progress":
        return "in_progress"
    return "queued"


class GitHubProvider(Provider):
    kind = "GitHub"
    max_body_length = 65536

    def __init__(self, remote: Remote, client: httpx.AsyncClient | None = None) -> None:
        self.remote = remote
        api = "https://api.github.com" if remote.host == "github.com" else f"https://{remote.host}/api/v3"
        self.api = f"{api}/repos/{remote.owner}/{remote.repo}"
        self.repo_url = f"https://{remote.host}/{remote.owner}/{remote.repo}"
        self.http = Http(
            GitHubAuth(remote.host),
            {"Accept": "application/vnd.github+json", "X-GitHub-Api-Version": "2022-11-28"},
            client,
        )

    async def _get(self, path: str, **params: Any) -> Any:
        return await self.http.json("GET", f"{self.api}{path}", params=params or None)

    def _pr(self, d: dict[str, Any]) -> PullRequest:
        state = "merged" if d.get("merged_at") else d["state"]
        return PullRequest(
            id=d["number"],
            title=d["title"],
            body=d.get("body") or "",
            state=state,
            draft=bool(d.get("draft")),
            source_branch=d["head"]["ref"],
            target_branch=d["base"]["ref"],
            author=(d.get("user") or {}).get("login", ""),
            url=d["html_url"],
            head_sha=d["head"]["sha"],
        )

    def _run(self, d: dict[str, Any]) -> PipelineRun:
        return PipelineRun(
            id=d["id"],
            name=d.get("name") or d.get("display_title") or str(d["id"]),
            status=_status(d.get("status")),
            result=_result(d.get("conclusion")),
            branch=d.get("head_branch") or "",
            commit=d.get("head_sha"),
            url=d["html_url"],
            started=d.get("run_started_at") or d.get("created_at"),
            finished=d.get("updated_at") if d.get("status") == "completed" else None,
            event=d.get("event"),
        )

    async def default_branch(self) -> str:
        return (await self._get(""))["default_branch"]

    async def find_open_pr(self, branch: str) -> PullRequest | None:
        prs = await self._get("/pulls", head=f"{self.remote.owner}:{branch}", state="open")
        return self._pr(prs[0]) if prs else None

    async def get_pr(self, pr_id: int) -> PullRequest:
        return self._pr(await self._get(f"/pulls/{pr_id}"))

    async def create_pr(self, title: str, body: str, source: str, target: str, draft: bool) -> PullRequest:
        payload = {"title": title, "body": body, "head": source, "base": target, "draft": draft}
        return self._pr(await self.http.json("POST", f"{self.api}/pulls", json=payload))

    async def update_pr(self, pr_id: int, title: str | None, body: str | None) -> PullRequest:
        payload = {k: v for k, v in (("title", title), ("body", body)) if v is not None}
        return self._pr(await self.http.json("PATCH", f"{self.api}/pulls/{pr_id}", json=payload))

    async def list_runs(self, branch: str | None, pr: PullRequest | None, limit: int) -> list[PipelineRun]:
        params: dict[str, Any] = {"per_page": limit}
        if pr is not None and pr.head_sha:
            params["head_sha"] = pr.head_sha
        elif branch:
            params["branch"] = branch
        data = await self._get("/actions/runs", **params)
        return [self._run(r) for r in data.get("workflow_runs", [])[:limit]]

    async def run_report(self, run_id: int, log_lines: int) -> RunReport:
        run = self._run(await self._get(f"/actions/runs/{run_id}"))
        jobs_data = (await self._get(f"/actions/runs/{run_id}/jobs", filter="latest", per_page=100)).get("jobs", [])
        jobs = [
            JobResult(j["name"], _status(j.get("status")), _result(j.get("conclusion")), j.get("html_url"))
            for j in jobs_data
        ]
        failures: list[FailedStep] = []
        for j in jobs_data:
            if _result(j.get("conclusion")) != "failure":
                continue
            steps = [s["name"] for s in j.get("steps") or [] if _result(s.get("conclusion")) == "failure"]
            errors = await self._annotations(j["id"])
            log = await self._job_log(j["id"], log_lines)
            failures.append(FailedStep(j["name"], ", ".join(steps) or "(job)", errors, log))
        return RunReport(run, jobs, failures)

    async def _annotations(self, job_id: int) -> list[str]:
        # A job's id is also its check-run id; failure annotations carry the error messages.
        try:
            notes = await self._get(f"/check-runs/{job_id}/annotations", per_page=50)
        except DevOpsError:
            return []
        return [n["message"] for n in notes if n.get("annotation_level") == "failure"][:20]

    async def _job_log(self, job_id: int, lines: int) -> str | None:
        if lines <= 0:
            return None
        try:
            resp = await self.http.request("GET", f"{self.api}/actions/jobs/{job_id}/logs")
        except DevOpsError:
            return None  # logs expire or are not ready yet; the report is still useful without them
        return failure_excerpt(_TIMESTAMP.sub("", resp.text), lines)

    async def pr_checks(self, pr: PullRequest) -> list[Check]:
        sha = pr.head_sha or (await self.get_pr(pr.id)).head_sha
        runs = (await self._get(f"/commits/{sha}/check-runs", per_page=100)).get("check_runs", [])
        checks = [
            Check(r["name"], _status(r.get("status")), _result(r.get("conclusion")), None, r.get("html_url"))
            for r in runs
        ]
        statuses = (await self._get(f"/commits/{sha}/status")).get("statuses", [])
        for s in statuses:
            done = s["state"] != "pending"
            checks.append(
                Check(s["context"], "completed" if done else "in_progress", _result(s["state"]) if done else None, None, s.get("target_url"))
            )
        return checks
